aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -101,6 +101,7 @@ TORCH_META_FUNC(gather)
     at::assert_no_partial_overlap(result, index);
   }
 
+  if (index.numel() == 0) return;
```

Comment:
I think this should be after `TORCH_CHECK`.

If this is here, I think this would allow index with any scalar_type if `index.numel()` is `0`